RFC-0026-logging-system.md

```diff
@@ -0,0 +1,103 @@
+# PyTorch Logging System
+
+## **Summary**
+Create a message logging system for PyTorch with the following requirements:
+
+* All errors, warnings, and other messages generated by PyTorch should be
+  emitted using the the logging system API
+
+* The APIs for emitting messages and changing settings should all be consistent
+  between C++ and Python
+
+* Offer different message severity levels, including at least the following:
+
+  - **Info**: Emits a message without creating a warning or error. By default,
+    this gets printed to stdout
+
+  - **Warning**: Emits a message as a warning. By default, this will turn into
+    a Python warning
+
+  - **Error**: Emits a message as an error. By default, this will turn into
+    a Python error
+
+  - TODO: Should we also have a **Fatal** severity for integration with
+    Meta's internal logging system? A fatal message terminates the program
+
+* Offer different classes of messages, including at least the following:
```

Comment:
Should we include all of the classic python error/warnings here (TypeError, ValueError, NotImplementedError, DeprecationWarning, etc) ?
Note that we already use all of these internally

Reply:
That may be the right idea.

However, I don't think I completely understand what @mruberry meant by "classes of messages" in the description of https://github.com/pytorch/pytorch/issues/72948. I was thinking that class and severity would be independent of each other--every message class can have any severity level. But if, for example, we say that `TypeError` is one of the message classes, then the only possible severity level for it would be Error

Reply:
I might have been overthinking this. Maybe it's actually totally fine for a particular message class to only have one severity level that it can be invoked with. In that case, the definition of "message class" could include each of the classic Python errors and warnings, plus any others that we would like to add, like nondeterministic/beta/prototype messages.

Still, I'm not completely sure what @mruberry had in mind

Reply:
Sorry for my confusion, I'm pretty sure I now understand what a "message class" is.

For each severity level, several different message classes can be used. A given message class is only used for one particular severity level. (So my thought that severity and class are independent was wrong.)

For example, with the error severity level in C++, we have the error class `c10::Error` and all of its subclasses defined [here](https://github.com/pytorch/pytorch/blob/72e4aab74b927c1ba5c3963cb17b4c0dce6e56bf/c10/util/Exception.h#L195-L236). When a C++ error propagates to Python, it turns into a corresponding Python error class. For instance `c10::Error` becomes `RuntimeError`, `c10::IndexError` becomes `IndexError`, etc.

At the moment, it looks like only one warning class exists in C++ (although I think it's not actually implemented as a C++ class), and it always propagates to the same Python class, `Warning`. I think one of the goals of this message logging system is to create different warning classes in C++ that propagate to a corresponding Python warning class--much like how error classes currently work--and to also do the same for info severity messages.

So I think the answer to @albanD's question "Should we include all of the classic python error/warnings here?" is probably yes. I'll update the requirements to reflect all of this